test/unit/eigensolver/mc/test_reduction_to_band.cpp

```diff
@@ -0,0 +1,320 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2019, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+
+#include "dlaf/eigensolver/reduction_to_band.h"
+
+#include <cmath>
+
+#include <gtest/gtest.h>
+
+#include "dlaf/common/index2d.h"
+#include "dlaf/common/pipeline.h"
+#include "dlaf/communication/communicator.h"
+#include "dlaf/communication/communicator_grid.h"
+#include "dlaf/communication/functions_sync.h"
+#include "dlaf/communication/sync/broadcast.h"
+#include "dlaf/lapack/tile.h"  // Just for lapack.hh workaround
+#include "dlaf/matrix/copy.h"
+#include "dlaf/matrix/matrix.h"
+#include "dlaf/memory/memory_view.h"
+#include "dlaf/types.h"
+#include "dlaf/util_matrix.h"
+
+#include "dlaf_test/comm_grids/grids_6_ranks.h"
+#include "dlaf_test/matrix/matrix_local.h"
+#include "dlaf_test/matrix/util_matrix.h"
+#include "dlaf_test/matrix/util_tile.h"
+#include "dlaf_test/util_types.h"
+
+using namespace dlaf;
+using namespace dlaf::test;
+using namespace dlaf::comm;
+using namespace dlaf::matrix;
+using namespace dlaf::matrix::test;
+using namespace testing;
+
+::testing::Environment* const comm_grids_env =
+    ::testing::AddGlobalTestEnvironment(new dlaf::test::CommunicatorGrid6RanksEnvironment);
+
+template <typename Type>
+class ReductionToBandTest : public ::testing::Test {
+public:
+  const std::vector<CommunicatorGrid>& commGrids() {
+    return dlaf::test::comm_grids;
+  }
+};
+
+TYPED_TEST_SUITE(ReductionToBandTest, dlaf::test::MatrixElementTypes);
+
+const std::vector<LocalElementSize> square_sizes{{3, 3}, {13, 13}, {12, 12}, {24, 24}};
+const std::vector<TileElementSize> square_block_sizes{{3, 3}};
+
+template <class T>
+MatrixLocal<T> makeLocal(const Matrix<const T, Device::CPU>& matrix) {
+  return {matrix.size(), matrix.distribution().blockSize()};
+}
+
+template <class T, Device device>  // TODO add tile_selector predicate
+void all_gather(Matrix<const T, device>& source, MatrixLocal<T>& dest,
+                comm::CommunicatorGrid comm_grid) {
+  const auto& dist_source = source.distribution();
+  const auto rank = dist_source.rankIndex();
+
+  for (const auto& ij_tile : iterate_range2d(dist_source.nrTiles())) {
+    const auto owner = dist_source.rankGlobalTile(ij_tile);
+
+    auto& dest_tile = dest.tile(ij_tile);
+
+    if (owner == rank) {
+      const auto& source_tile = source.read(ij_tile).get();
+      comm::sync::broadcast::send(comm_grid.fullCommunicator(), source_tile);
+      copy(source_tile, dest_tile);
```

Comment:
`copyTile`?

Reply:
`copyTile` is an internal machinery used by `copy`.